### PR TITLE
[PROPOSAL] Use Google Test library

### PR DESCRIPTION
**Pull request by NullSoldier**
I want to propose using google-test as the official testing framework here before I do any more work on this branch. *(disclaimer: I'm a bit rough on my C++, but I'm fixing that)*

I think there are some clear benefits, especially for a library which is so heavily used and supported. This is a really rough proposal. I would finish importing over the rest of the tests piecemeal before killing the old tests and renaming the old directory.

*Organization*
A testing framework will organize tests into clearly separate test cases. Looking through the tests right now, it's difficult to see what is actually tested, and what is not tested. google-test will provide a more clear organization for tests that make writing tests easier (because you know where they should go) and finding tests (know what is or is not tested)

*Ease of Use*
Google has plenty of resources on how to use google-test, as well as google-mock. If anyone needs to find out how to write a unit test or take advantage of useful google-test features they can find all the resources they need online. google-test has a reasonably sized support community

*Standardization*
Google-test is an XUnit test framework which means we can pipe the output of the test runner into tons of tools that are already made to consume XUnit results. 

And all the other standard benefits...
- Automatic test discovery (scalability)
- Nice test output from the test runner
- Create custom semantic asserts
- Plugins for gtest to parallelize test runs

screen shot 2016-08-11 at 12 34 35 am

### CODE REVIEW

File: exampletests/asserts.h [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+        uint16_t fragExpected = htons(expected[i]);
+
+        if(fragActual != fragExpected)
+            return testing::AssertionFailure()
+                << "ip fragment " << (i+1) << " does not match "
+                << fragActual << " != " << fragExpected;
+    }
+
+    return testing::AssertionSuccess();
+}
```
Comment (NullSoldier) [inline exampletests/asserts.h]: A small example of what you can do with semantic asserts. I would probably change this to  have better error messages and even check byte order.

File: exampletests/test_address.cpp [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+TEST(Address, Clear)
+{
+    Address address( 127, 0, 0, 1, 1000 );
+    address.Clear();
+
+    ASSERT_FALSE( address.IsValid() );
+    ASSERT_EQ( address.GetType(), ADDRESS_NONE );
+    ASSERT_EQ( address.GetPort(), 0 );
+    // ASSERT_EQ( address.GetAddress4(), (uint32_t)0x0 );
+}
```
Comment (NullSoldier) [inline exampletests/test_address.cpp]: This is actually a bug I found while porting these tests. The underlying ipv4_address member isn't cleared?

**Comment by gafferongames** [comment]
This is interesting. Lets talk about it:

a) Why google test? Is it the defacto standard?

b) What other testing frameworks are available?

c) Is it widely available on Windows, MacOS and Linux?

d) Can it be installed nicely with brew on MacOSX?

e) Are there packages for google test on Ubuntu/Debian?

f) How is google test installed on Windows?

g) Can I run the tests standalone? My current workflow while developing tests is to run "premake5 test" locally while adding new tests? Can this be replicated with google test?

h) Does google test require "modern" C++ features?

thanks

**Comment by onqtam** [comment]
@gafferongames I would suggest [Catch](https://github.com/philsquared/Catch) as it is just a single header with a nice interface instead of google test. Alternatively you may choose [doctest](https://github.com/onqtam/doctest) which is my reimplementation of catch for compile times (shameless advertisement :D)

**Comment by grahamreeds** [comment]
a) Yes it is. Mainly because Google is behind it, but also because it is
cross platform, has lots of users so any problems you have will most likely
to be found on SO and because it comes with a mocking library that follows
a similar design. If you go for my favourite, UnitTest++, or Catch as
another commentator has mentioned you then have to pick a mocking library
(Hippo Mocks) which is written by another person who has their own views on
what makes a good library.

b) Lots. I am biased towards UnitTest++ and the aforementioned Catch. None
come with a mocking library as standard and you do need to mock out
connections etc.

c) Yes. Most C++ unit testing frameworks are.

d) I would assume so but I don't have a Mac.

e) Probably but I use UnitTest++ which is cmake based.

f) Unsure. Someone else needs to answer this one but I assume you can add
them easily as a submodule from github.

g) I can't answer that one but in the case of UnitTest++ the answer is no,
but as 200 tests run in 0.2 seconds on my U14_04 VM with 4 cores and 2gb
ram running them after every compile isn't a hardship.

h) No. But since we (C++) lack reflection most use macro hackery to
register tests.

Microsoft use UnitTest++ for their cross platform stuff. Casablanca and the
GSL are the two highest profile cases. UnitTest was developed by two Indy
game devs.

GR

**Comment by NullSoldier** [comment]
Sorry I haven't been able to respond I'm itching to jump into the conversation. I'll have a response later after work today. This type of collaboration / discussion is why I enjoy open source contributions. I'm also doing this to become familiar with the library.

**Comment by NullSoldier** [comment]
I think a decent amount of the questions were answered well by Graham so I'll just answer the ones I think I may have different answers too.

> d) Can it be installed nicely with brew on MacOSX?
> e) Are there packages for google test on Ubuntu/Debian?
> f) How is google test installed on Windows?

After doing some asking around, the recommended / common way would be to ship the code with the project in a vendor or external folder similar to what I've done in the example just not with a submodule. There would be no installation needed.

> g) Can I run the tests standalone? My current workflow while developing tests is to run "premake5 test" locally while adding new tests? Can this be replicated with google test?

Your workflow will be the exact same, "premake5 test" would build the test and project and run it as a post build command the exact same as you have now.

> What is the value of xtest results? How would I used them? Do they integrate with Travis-CI? What does that look like?

This one is a bit more general. Since google-test has a lot of support, no matter which editor you use, you can probably find a google-test runner plugin that displays test runs and test results in the editor. Since the project is very small, tools around visualizing test results don't matter as much. Later, we can bring in tools around processing the XUnit results. One useful example would be "show the top 5 slowest tests". Having Travis build a nice HTML artifact which displays the test failures for PRs would be very helpful.

There's so many things in this domain that really it's more about opening doors later for when you want to pull in tools to process the standardized test results.

I recommend google-test because of how widely supported it is. It's as easy as most unit testing frameworks, but with more advanced features for when you need it. Granted, I would be okay with any framework that can at least provide an organizational structure for writing more concise and clear tests.

**Comment by grahamreeds** [comment]
Another cool thing about using a framework you can easily link Travis to a
coverage tool and get feedback on how much of your code the tests cover.

This is something I mean to do for UnitTest++, and I link to various online
sources showing how in an issue in the UT repo.

GR

**Comment by gafferongames** [comment]
Thanks for this suggestion. I've thought about it over the last day and I think, seeing as google test is itself larger than the library it is being included to test, and on MacOSX needs to be bundled as a library + headers inside my library zip to ship to users, I think it is overkill (for now).

In the future I might reconsider, but for now I think the very minimal testing functionality (which is really, more about *functional* tests than exhaustive unit testing of every single function...), is good enough for now.

Thanks for your suggestion. I do think it would still be a useful exercise for you to run the library through google test on your side if you would like to do that as a learning exercise. I encourage you fork and do that if you would like, but I don't think I can accept it back in the main branch because it is such a large dependency. I want to make it as simple as possible for people to build and run tests on all platforms, without adding additional dependencies for the library (mbedtls and libsodium are a lot already!)

thanks!

**Comment by gafferongames** [comment]
> Another cool thing about using a framework you can easily link Travis to a
> coverage tool and get feedback on how much of your code the tests cover.

I think this is a good idea for later on. Close to 1.0 I think it would be good to get more extensive testing in (in a Unit Testing style) as the API finalizes. Right now I think it would provide too much rigidity to the API if I were to change it (I will... I'm only at 0.3)

cheers

**Pull request closed without merging**